File: built-in plugins dsp/mb-strings-violin.cpp
/**
 * MB Solo Violin
 * Category : instrument
 * Type     : strings
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Expressive solo violin
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_STRINGS_VIOLIN_H
#define MB_STRINGS_VIOLIN_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbStringsViolin : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-strings-violin";
    static constexpr const char* PLUGIN_NAME    = "MB Solo Violin";
    static constexpr const char* PLUGIN_TYPE    = "strings";
    static constexpr const char* PLUGIN_CATEGORY = "instrument";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float vibrato = 0.5f;  // range [0, 1]
    float volume = 0.8f;  // range [0, 1]
    };

    MbStringsViolin() = default;
    ~MbStringsViolin() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.vibrato = std::clamp(params.vibrato, 0f, 1f);
        params.volume = std::clamp(params.volume, 0f, 1f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Solo Violin
        return input;
    }
};

#endif // MB_STRINGS_VIOLIN_H

File: built-in plugins dsp/mb-strings-violin.ts
import type { PluginDefinition } from "../server/services/pluginHostService";

const MbStringsViolinPlugin: PluginDefinition = {
  id: "mb-strings-violin",
  slug: "mb-strings-violin",
  name: "MB Solo Violin",
  category: "instrument",
  type: "strings",
  version: "1.0.0",
  description: "Expressive solo violin",
  author: "Max Booster",
  grade: "A",
  oscillators: [
    { type: "sawtooth", detune: 0, gain: 0.6 },
    { type: "triangle", detune: 3, gain: 0.4 },
  ],
  envelope: { attack: 0.05, decay: 0.3, sustain: 0.9, release: 0.4 },
  parameters: [
    {
      id: "vibrato",
      name: "Vibrato",
      type: "float",
      defaultValue: 0.5,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
    {
      id: "volume",
      name: "Volume",
      type: "float",
      defaultValue: 0.8,
      minValue: 0,
      maxValue: 1,
      automatable: true,
    },
  ],
  defaultPreset: { vibrato: 0.5, volume: 0.8 },
};

export default MbStringsViolinPlugin;
